### DISCUSSION TITLE
Results object plot method

### DISCUSSION THREAD

**Discussion by jadrian21** [ultralytics/ultralytics#9948] [Docs]
Hi, How can I use the plot method and not plot some boxes due to them being implausible detections: for example, a small bounding box
for a person in a room (actually recognising a toy). I can rule out counting the detection as valid by ignoring bounding boxes by height or width but I don't know how to not plot this box when using the plot method.

**Comment by glenn-jocher** [comment]
Hello!

Great question! To avoid plotting certain bounding boxes based on their size, you can modify the `Results` object directly before calling the `plot()` method. Here's a concise way to filter out boxes by their dimensions before plotting:

```python
from ultralytics import YOLO

# Load model and run prediction
model = YOLO('yolov8n.pt')
results = model('your_image.jpg')

# Define your size criteria (for example, width < 50 or height < 50)
min_width = 50
min_height = 50
filtered_boxes = []

# Filter boxes
for box in results[0].boxes.xywh:
    x, y, w, h = box
    if w > min_width and h > min_height:
        filtered_boxes.append(box)

# Update results with filtered boxes
results[0].boxes.xywh = filtered_boxes

# Plot results with filtered boxes
annotated_img = results[0].plot()
```

This snippet filters out detections where the bounding box width or height is smaller than the specified minimums, and then plots the remaining detections. Remember to adjust `min_width` and `min_height` according to your needs. Happy coding! 😊

**Reply by jadrian21** [reply]
Hi that code gave the following error:

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[36], [line 19](vscode-notebook-cell:?execution_count=36&line=19)
     [16](vscode-notebook-cell:?execution_count=36&line=16)         filtered_boxes.append(box)
     [18](vscode-notebook-cell:?execution_count=36&line=18) # Update results with filtered boxes
---> [19](vscode-notebook-cell:?execution_count=36&line=19) results[0].boxes.xywh = filtered_boxes
     [21](vscode-notebook-cell:?execution_count=36&line=21) # Plot results with filtered boxes
     [22](vscode-notebook-cell:?execution_count=36&line=22) annotated_img = results[0].plot()

AttributeError: property 'xywh' of 'Boxes' object has no setter

**Reply by glenn-jocher** [reply]
@jadrian21 hi there! 😊 It looks like you encountered an error because the `xywh` attribute of a `Boxes` object is a property without a setter method, meaning we can't directly assign a new value to it like `results[0].boxes.xywh = filtered_boxes`.

Instead, here’s a different approach to filter and plot the boxes. You can manually draw on the image using OpenCV right after performing the size filtering:

```python
import cv2
from ultralytics import YOLO

# Load model and run prediction
model = YOLO('yolov8n.pt')
results = model('your_image.jpg')

# Initial image for drawing
img = results.imgs[0].copy()

# Filter and plot
min_width = 50
min_height = 50
for *xyxy, conf, cls in results.xyxy[0]:  # results.xyxy[0] is the detections tensor
    x1, y1, x2, y2 = map(int, xyxy)
    if (x2 - x1) > min_width and (y2 - y1) > min_height:  # Filter based on size
        color = (255, 0, 0)  # Color for the bounding box (BGR)
        cv2.rectangle(img, (x1, y1), (x2, y2), color, thickness=2)  # Draw

# Display your filtered and annotated image
cv2.imshow('Filtered Detections', img)
cv2.waitKey(0)  # Press any key to close the image window
```
This method relies on manually drawing the filtered boxes on the original image using OpenCV. Adjust `min_width` and `min_height` to suit your needs. Have fun coding!

**Reply by jadrian21** [reply]
Hi Glenn, that code gave this error:

image 1/1 [c:\Users\jadrian\AnacondaProjects\OpenCV](file:///C:/Users/jadrian/AnacondaProjects/OpenCV) Files\BedEdge.png: 640x640 1 person, 1 couch, 1 book, 5.0ms
Speed: 2.0ms preprocess, 5.0ms inference, 1.0ms postprocess per image at shape (1, 3, 640, 640)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[17], [line 9](vscode-notebook-cell:?execution_count=17&line=9)
      [6](vscode-notebook-cell:?execution_count=17&line=6) results = model('BedEdge.png')
      [8](vscode-notebook-cell:?execution_count=17&line=8) # Initial image for drawing
----> [9](vscode-notebook-cell:?execution_count=17&line=9) img = results.imgs[0].copy()
     [11](vscode-notebook-cell:?execution_count=17&line=11) # Filter and plot
     [12](vscode-notebook-cell:?execution_count=17&line=12) min_width = 50

AttributeError: 'list' object has no attribute 'imgs'

**Reply by glenn-jocher** [reply]
Hey there! 🙋‍♂️ It looks like there's a slight mix-up in accessing the image data from the `Results` object. The correct way to get the original image after predictions would be to use `results.imgs[0]` if `imgs` were a direct attribute of `Results`. However, it appears this attribute might not exist directly as anticipated. 

An alternative approach to manipulate or visualize the results is to work directly with the image you passed to the model. You could use OpenCV for reading the image and then draw the filtered bounding boxes on it. Here’s a quick example on how you could adjust the workflow:

```python
import cv2
from ultralytics import YOLO

# Load the model and run prediction
model = YOLO('yolov8n.pt')
results = model('BedEdge.png')

# Load the original image with OpenCV
img = cv2.imread('BedEdge.png')

# Continue with the filtering and drawing as needed

# [Your filtering and drawing code here]

# Display the result
cv2.imshow('Filtered Detections', img)
cv2.waitKey(0)
cv2.destroyAllWindows()
```

This way, you're directly controlling the image manipulation using OpenCV, ensuring you won't hit the same attribute error. Hope this helps, and happy coding! 😊
